File: trunk/FRC-3128-2014/src/frc3128/Util/Connection/RobotCommand.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

package frc3128.Util.Connection;

import frc3128.Util.DebugLog;
import java.util.Vector;

/**
 *
 * @author jamie
 */
public class RobotCommand 
{
    //this is an unsigned int in the c++ program
    //but we need to round up here
    //java, folks
    int _return_id;
    
    byte _opcode;
    
    Vector _shorts;
    
    String _extraString;
    
    public RobotCommand(int return_id, byte opcode)
    {
        _opcode = opcode;
    }
    
    public RobotCommand(int return_id, byte opcode, Vector shorts)
    {
        _opcode = opcode;

        _shorts = shorts;
    }

    public RobotCommand(int return_id, byte opcode, Vector shorts, String extraString)
    {
        _opcode = opcode;

        _shorts = shorts;

        _extraString = extraString;
    }

    public RobotCommand(int return_id, byte opcode, String extraString)
    {
        _opcode = opcode;

        _extraString = extraString;
    }

    public String toString()
    {
        String string = "Robot Command dump:---------------\n";
        
        string += "Return ID: " + _return_id + "\n";

        string += "Opcode: " + Integer.toHexString(_opcode) + "\n";

        //print out each short
        if(_shorts != null)
        {
            string += "Shorts: ";
            for(int counter = _shorts.size(); counter >= 1; --counter)
            {
                string += _shorts.elementAt(counter) + " ";
            }
            string += "\n";
        }

        string += "Extra String: " + _extraString + "\n";
        return string;
    }

        //Parsing code originally written in c++, where it was a lot... better
    
    public static RobotCommand factory(byte[] bytes)
    {
        //stone knives and bearskins...
        Vector currentBytePtr = new Vector();
        for(int counter = 0; counter < bytes.length; counter++)
        {
            currentBytePtr.addElement(Byte.valueOf(bytes[counter]));
        }
        
        return factory(currentBytePtr);
    }

    public static RobotCommand factory(Vector currentBytePtr)
    {
        int return_id = parseReturnID(currentBytePtr);
        
        ++iterator;
        
        byte opcode = parseOpcode(currentBytePtr);
        Vector shorts = parseShorts(currentBytePtr);
        String extraString = parseString(currentBytePtr);

        iterator = 0;
        return new RobotCommand(return_id, opcode, shorts, extraString);
    }

        //needs to be class scope so that it can keep its state after a function call
        static int iterator = 0;

    static byte parseOpcode(Vector currentBytePtr)
    {
        // Get opcode.
        ++iterator;
        byte opcode = ((Byte)currentBytePtr.elementAt(iterator)).byteValue();
        ++iterator;

        if(((Byte)currentBytePtr.elementAt(iterator)).byteValue() != SOSProtocol.END_OPCODE)
        {
                // Bail and return empty value.
                DebugLog.log(DebugLog.LVL_ERROR, null, "RobotCommand: the end of the opcode command is incorrect\nShould be:" + SOSProtocol.END_OPCODE + " Was: " + ((Byte)currentBytePtr.elementAt(iterator)).byteValue());
                return 0x0;
        }

        return opcode;
    }
    
    static int parseReturnID(Vector currentBytePtr)
    {
        if(((Byte)currentBytePtr.elementAt(iterator)).byteValue() != SOSProtocol.START_TRANSMISSION)
        {
                // Bail and return empty value.
            DebugLog.log(DebugLog.LVL_ERROR, null, "RobotCommand: the header of the command is incorrect\nShould be:" + SOSProtocol.START_TRANSMISSION + " Was: " + currentBytePtr.elementAt(iterator));
            return 0x0;
        }

	// Have start of command.  Begin parsing.


	++iterator;
	//should now equal START_ID
	++iterator;
	//should now be the first byte of the ASCII-encoded id

        //max of 10 ASCII digit short
        char[] returnBytesStorage = new char[10];
        int returnBytesIterator = 0;

        while(((Byte)currentBytePtr.elementAt(++iterator)).byteValue() != SOSProtocol.END_SHORT)
        {
                returnBytesStorage[returnBytesIterator] = (char)(((Byte)currentBytePtr.elementAt(iterator)).byteValue());
                ++returnBytesIterator;
        }

        return Integer.valueOf(String.valueOf(returnBytesStorage)).intValue();
    }

static Vector parseShorts(Vector currentBytePtr)
{
	if(((Byte)currentBytePtr.elementAt(iterator)).byteValue() == SOSProtocol.START_SHORT_TRANSMISSION)
	{
		Vector shortVector = new Vector();

		 // Loop until we don't recieve the start of a short
		while(((Byte)currentBytePtr.elementAt(iterator)).byteValue() == SOSProtocol.START_SHORT_TRANSMISSION)
		{
                    //max of 10 ASCII digit short
			char[] shortBytesStorage = new char[10];
                        int shortBytesIterator = 0;
                        
			while(((Byte)currentBytePtr.elementAt(++iterator)).byteValue() != SOSProtocol.END_SHORT)
			{
				shortBytesStorage[shortBytesIterator] = (char)(((Byte)currentBytePtr.elementAt(iterator)).byteValue());
                                ++shortBytesIterator;
			}
                        
			shortVector.addElement(Integer.valueOf(String.valueOf(shortBytesStorage)));
			++iterator;
		}

		return shortVector;
	}
	else
	{
		return null;
	}

}

static String parseString(Vector currentBytePtr)
{
	if(((Byte)currentBytePtr.elementAt(iterator)).byteValue() == SOSProtocol.START_STRING_TRANSMISSION)
	{
		String stringstream = new String();
		while(((Byte)currentBytePtr.elementAt(++iterator)).byteValue()!= SOSProtocol.END_STRING)
		{
        		stringstream += ((Byte)currentBytePtr.elementAt(iterator)).byteValue();
		}

		return stringstream;
	}
        
        return null;
}

}

File: trunk/libSOS/client/RobotCommand.c
#include "RobotCommand.h"
#include "SOSProtocol.h"

#include <stdio.h>
#include <stdlib.h>
#include <assert.h>

#define member_size(type, member) sizeof(((type *)0)->member)

in_port_t parse_return_id(char * currentBytePtr, int * iterator)
{
	if(*(currentBytePtr + *(iterator)) != START_TRANSMISSION)
	{
		// Bail and return empty value.
		printf("RobotCommand: the header of the command is incorrect\nShould be: %x Was: %x\n", START_TRANSMISSION, *(currentBytePtr + *(iterator)));
		return 0x0;
	}

	// Have start of command.  Begin parsing.


	++*(iterator);
	//should now equal START_ID
	++*(iterator);
	//should now be the first byte of the ASCII-encoded id

	char* IDStorage = malloc(11);
	int IDStorageIterator = 0;
	
	while(*(currentBytePtr + (++*(iterator))) != END_ID)
	{
		*(IDStorage + IDStorageIterator) = *(currentBytePtr + *(iterator));
		++IDStorageIterator;
	}
	
	//insert a null to stop atoi from
	//reading into unknown memory
	IDStorage[10] = 0x0;
	
	//do this here so we can free IDStorage
	in_port_t return_id = atoi(IDStorage);

	free(IDStorage);
	
	//should now equal END_ID
	return return_id;
}
char parse_opcode(char * currentBytePtr, int * iterator)
{
	// Get opcode.
	++*(iterator);
	char opcode = *(currentBytePtr + *(iterator));
	++*(iterator);

	if(*(currentBytePtr + *(iterator)) != END_OPCODE)
	{
		// Bail and return empty value.
		printf("RobotCommand: the end of the opcode command is incorrect\nShould be: %x Was: %x\n", END_OPCODE, *(currentBytePtr + *(iterator)));
		return 0x0;
	}
	return opcode;
}

//reads as many shorts as possible from the command
//if it can't find any, it returns NULL
short * parse_shorts(char * currentBytePtr, int * iterator, unsigned int * shorts_len)
{
	*(shorts_len) = 0;

	if(*(currentBytePtr + *(iterator)) == START_SHORT_TRANSMISSION)
	{
		short * shortVector = NULL;
		*(shorts_len) = 0;
		
		 // Loop until we don't receive the start of a short
		while(*(currentBytePtr + *(iterator)) == START_SHORT_TRANSMISSION)
		{
			char shortBytesStorage[11];
			unsigned int shortBytesIterator = 0;
			
			while(*(currentBytePtr + (++(*(iterator)))) != END_SHORT)
			{
				*(shortBytesStorage + shortBytesIterator) = *(currentBytePtr + *(iterator));
			}
			
			//increase size of shortVector by 1
			shortVector = (short *) realloc(shortVector, (*(shorts_len) + 1) * sizeof(short));
			++shorts_len;
			
			//provide an endstop for atoi
			shortBytesStorage[10] = 0x0;
			
			*(shortVector + *(shorts_len)) = atoi(shortBytesStorage);
			++(*(iterator));
		}

		return shortVector;
	}
	else
	{
		return NULL;
	}

}

//tries to read a string from the bytes given
//returns null if it fails
char* parse_string(char * currentBytePtr, int * iterator)
{
	if(*(currentBytePtr + *(iterator)) == START_STRING_TRANSMISSION)
	{
		char * string;
		int counter = 0;
		while(*(currentBytePtr + (++(*(iterator)))) != END_STRING)
		{
			*(string + counter) = *(currentBytePtr + *(iterator));
			++counter;
		}

		return string;
	}
	else
	{
		return NULL;
	}

}

RobotCommand * factory(char * currentBytePtr)
{
	int 									iterator = 0;
	in_port_t								return_id;
	char 									opcode;
	short *			 						shorts = NULL;
	unsigned int							shorts_len = 0;
	char* 									string = NULL;


	return_id = parse_return_id(currentBytePtr, &iterator);

	++iterator;
	opcode = parse_opcode(currentBytePtr, &iterator);

	// Move into shorts if any.
	++iterator;
	// should NOW equal END_TRANSMISSION or START_SHORT_TRANSMISSION

	//returns an empty optional if there isn't a short
	shorts = parse_shorts(currentBytePtr, &iterator, &shorts_len);
	
	string = parse_string(currentBytePtr, &iterator);
	
	//maybe someday we'll find a way to utilise our variable-length short parser
	assert(shorts_len <= sizeof(member_size(RobotCommand, shorts)));
}